apiserver/facades/client/modelupgrader/controllerupgrader.go

```diff
@@ -0,0 +1,154 @@
+// back to a prior version.
+func (c *ControllerUpgraderAPI) AbortModelUpgrade(ctx context.Context, arg params.ModelParam) error {
+	return errors.NotSupportedf("abort model upgrade")
+}
+
+func (c *ControllerUpgraderAPI) canUpgrade(ctx context.Context, model names.ModelTag) error {
+	return nil
+}
+
+// UpgradeModel upgrades a controller and the model hosting the controller.
+// TODO(adisazhar123): support arg.dryRun.
+func (c *ControllerUpgraderAPI) UpgradeModel(ctx context.Context, arg params.UpgradeModelParams) (params.UpgradeModelResult, error) {
+	c.logger.Tracef(ctx, "UpgradeModel arg %#v", arg)
+	var result params.UpgradeModelResult
+
+	modelTag, err := names.ParseModelTag(arg.ModelTag)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	if err := c.canUpgrade(ctx, modelTag); err != nil {
+		return result, err
+	}
+	if err := c.check.ChangeAllowed(ctx); err != nil {
+		return result, errors.Trace(err)
+	}
+
+	return c.doUpgrade(ctx, arg)
+}
+
+// doUpgrade has the responsibility of delegating the upgrade to the service. It determines which func to invoke
+// by interrogating the values set in [params.UpgradeModelParams].
+// A post-processing step is performed to map the errors returned from the service to ones the existing API
+// conforms to.
+func (c *ControllerUpgraderAPI) doUpgrade(ctx context.Context, arg params.UpgradeModelParams) (params.UpgradeModelResult, error) {
+	var result params.UpgradeModelResult
+
+	var targetVersion = arg.TargetVersion
+	var err error
+
+	// Delegate it to the service depending on what arguments
+	// are supplied.
+	if arg.TargetVersion == semversion.Zero && arg.AgentStream == "" {
+		targetVersion, err = c.upgraderService.UpgradeController(ctx)
+	} else if arg.TargetVersion == semversion.Zero && arg.AgentStream != "" {
+		var stream modelagent.AgentStream
+		stream, err = modelagent.AgentStreamFromCoreAgentStream(agentbinary.AgentStream(arg.AgentStream))
+		if err == nil {
+			targetVersion, err = c.upgraderService.UpgradeControllerWithStream(ctx, stream)
+		}
+	} else if arg.TargetVersion != semversion.Zero && arg.AgentStream == "" {
+		err = c.upgraderService.UpgradeControllerToVersion(ctx, arg.TargetVersion)
+	} else if arg.TargetVersion != semversion.Zero && arg.AgentStream != "" {
+		var stream modelagent.AgentStream
+		stream, err = modelagent.AgentStreamFromCoreAgentStream(agentbinary.AgentStream(arg.AgentStream))
+		if err == nil {
+			err = c.upgraderService.UpgradeControllerToVersionAndStream(ctx, arg.TargetVersion, stream)
+		}
+	}
+	result.ChosenVersion = targetVersion
+
+	// Map the errors to respect what the existing API returns.
+	// We mirror as closely as possible to [UpgradeModel] func in [modelupgrader.ModelUpgraderAPI].
+	switch {
+	case errors.Is(err, controllerupgradererrors.MissingControllerBinaries):
+		result.Error = apiservererrors.ServerError(err)
+		return result, nil
+	case errors.HasType[controllerupgradererrors.ControllerUpgradeBlocker](err):
```

Comment:
If you use AsType you can get that actual value out and then just pluck the Reason string. The reason string is what you really want for the user.